File: exp4.py
from collections import defaultdict, deque

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.adj_matrix = [[0 for _ in range(vertices)] for _ in range(vertices)]  # Adjacency matrix
        self.adj_list = defaultdict(list)  # Adjacency list 

    # Add an edge for adjacency matrix and adjacency list
    def add_edge(self, u, v):
        # For adjacency matrix
        self.adj_matrix[u][v] = 1
        self.adj_matrix[v][u] = 1  # For undirected graph

        # For adjacency list
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)  # For undirected graph

    # Breadth-First Traversal using adjacency matrix
    def bft_matrix(self, start):
        visited = [False] * self.V
        queue = deque([start])
        visited[start] = True
 
        print("BFT using Adjacency Matrix:")
        while queue:
            node = queue.popleft()
            print(node, end=" ")

            # Traverse all vertices connected to the current node
            for i in range(self.V):
                if self.adj_matrix[node][i] == 1 and not visited[i]:
                    queue.append(i)
                    visited[i] = True
        print()

    # Breadth-First Traversal using adjacency list
    def bft_list(self, start):
        visited = [False] * self.V
        queue = deque([start])
        visited[start] = True

        print("BFT using Adjacency List:")
        while queue:
            node = queue.popleft()
            print(node, end=" ")

            # Traverse all neighbors of the current node
            for neighbor in self.adj_list[node]:
                if not visited[neighbor]:
                    queue.append(neighbor)
                    visited[neighbor] = True
        print()

    # Depth-First Traversal using adjacency matrix
    def dft_matrix(self, start):
        visited = [False] * self.V
        print("DFT using Adjacency Matrix:")
        self._dft_matrix_util(start, visited)
        print()

    def _dft_matrix_util(self, node, visited):
        visited[node] = True
        print(node, end=" ")

        # Traverse all vertices connected to the current node
        for i in range(self.V):
            if self.adj_matrix[node][i] == 1 and not visited[i]:
                self._dft_matrix_util(i, visited)

    # Depth-First Traversal using adjacency list
    def dft_list(self, start):
        visited = [False] * self.V
        print("DFT using Adjacency List:")
        self._dft_list_util(start, visited)
        print()

    def _dft_list_util(self, node, visited):
        visited[node] = True
        print(node, end=" ")

        # Traverse all neighbors of the current node
        for neighbor in self.adj_list[node]:
            if not visited[neighbor]:
                self._dft_list_util(neighbor, visited)


# Example Usage
if __name__ == "__main__":
    # Create a graph with 5 vertices
    g = Graph(5)
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    g.add_edge(1, 4)

    # Perform Breadth-First Traversal
    g.bft_matrix(0)  # Using adjacency matrix
    g.bft_list(0)    # Using adjacency list

    # Perform Depth-First Traversal
    g.dft_matrix(0)  # Using adjacency matrix
    g.dft_list(0)    # Using adjacency list

File: exp4.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100

// Queue structure for BFS
typedef struct {
    int items[MAX];
    int front, rear;
} Queue;

void initQueue(Queue* q) {
    q->front = q->rear = -1;
}

int isEmpty(Queue* q) {
    return q->front == -1;
}

void enqueue(Queue* q, int value) {
    if (q->rear == MAX - 1)
        return;
    if (isEmpty(q))
        q->front = 0;
    q->items[++q->rear] = value;
}

int dequeue(Queue* q) {
    if (isEmpty(q))
        return -1;
    int item = q->items[q->front];
    if (q->front >= q->rear)
        q->front = q->rear = -1;
    else
        q->front++;
    return item;
}

// -----------------------------
// Using Adjacency Matrix
// -----------------------------
void BFT_Matrix(int graph[MAX][MAX], int n, int start) {
    int visited[MAX] = {0}, i;
    Queue q;
    initQueue(&q);

    visited[start] = 1;
    enqueue(&q, start);

    printf("BFT using Adjacency Matrix: ");
    while (!isEmpty(&q)) {
        int v = dequeue(&q);
        printf("%d ", v);
        for (i = 0; i < n; i++) {
            if (graph[v][i] && !visited[i]) {
                visited[i] = 1;
                enqueue(&q, i);
            }
        }
    }
    printf("\n");
}

void DFT_Matrix_Helper(int graph[MAX][MAX], int n, int v, int visited[MAX]) {
    visited[v] = 1;
    printf("%d ", v);
    int i;
    for (i = 0; i < n; i++) {
        if (graph[v][i] && !visited[i]) {
            DFT_Matrix_Helper(graph, n, i, visited);
        }
    }
}

void DFT_Matrix(int graph[MAX][MAX], int n, int start) {
    int visited[MAX] = {0};
    printf("DFT using Adjacency Matrix: ");
    DFT_Matrix_Helper(graph, n, start, visited);
    printf("\n");
}

// -----------------------------
// Using Adjacency List
// -----------------------------
typedef struct Node {
    int vertex;
    struct Node* next;
} Node;

typedef struct Graph {
    int numVertices;
    Node* adjLists[MAX];
} Graph;

Graph* createGraph(int vertices) {
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    graph->numVertices = vertices;
    int i;
    for (i = 0; i < vertices; i++)
        graph->adjLists[i] = NULL;
    return graph;
}

void addEdge(Graph* graph, int src, int dest) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->vertex = dest;
    newNode->next = graph->adjLists[src];
    graph->adjLists[src] = newNode;

    newNode = (Node*)malloc(sizeof(Node));
    newNode->vertex = src;
    newNode->next = graph->adjLists[dest];
    graph->adjLists[dest] = newNode;
}

void BFT_List(Graph* graph, int start) {
    int visited[MAX] = {0};
    Queue q;
    initQueue(&q);

    visited[start] = 1;
    enqueue(&q, start);

    printf("BFT using Adjacency List: ");
    while (!isEmpty(&q)) {
        int v = dequeue(&q);
        printf("%d ", v);

        Node* temp = graph->adjLists[v];
        while (temp) {
            int adjVertex = temp->vertex;
            if (!visited[adjVertex]) {
                visited[adjVertex] = 1;
                enqueue(&q, adjVertex);
            }
            temp = temp->next;
        }
    }
    printf("\n");
}

void DFT_List_Helper(Graph* graph, int v, int visited[MAX]) {
    visited[v] = 1;
    printf("%d ", v);

    Node* temp = graph->adjLists[v];
    while (temp) {
        if (!visited[temp->vertex])
            DFT_List_Helper(graph, temp->vertex, visited);
        temp = temp->next;
    }
}

void DFT_List(Graph* graph, int start) {
    int visited[MAX] = {0};
    printf("DFT using Adjacency List: ");
    DFT_List_Helper(graph, start, visited);
    printf("\n");
}

// -----------------------------
// Main Program
// -----------------------------
int main() {
    int n = 5; // Number of vertices

    // Adjacency Matrix Example
    int matrix[MAX][MAX] = {0};
    matrix[0][1] = matrix[1][0] = 1;
    matrix[0][2] = matrix[2][0] = 1;
    matrix[1][3] = matrix[3][1] = 1;
    matrix[2][4] = matrix[4][2] = 1;

    BFT_Matrix(matrix, n, 0);
    DFT_Matrix(matrix, n, 0);

    // Adjacency List Example
    Graph* graph = createGraph(n);
    addEdge(graph, 0, 1);
    addEdge(graph, 0, 2);
    addEdge(graph, 1, 3);
    addEdge(graph, 1, 4);

    BFT_List(graph, 0);
    DFT_List(graph, 0);

    return 0;
}
